### PR TITLE
Created Server by Node.js

### PR DESCRIPTION
**Pull request by mr-olesha**

### CODE REVIEW

File: src/createServer.js [hunk 1]
```diff
@@ -1,3 +1,59 @@
-// Write code here
-// Also, you can create additional files in the src folder
-// and import (require) them here
+const http = require('http');
+const { convertToCase } = require('./convertToCase/convertToCase');
+
+const supportedCases = ['SNAKE', 'KEBAB', 'CAMEL', 'PASCAL', 'UPPER'];
```
Comment (BudnikOleksii) [inline src/createServer.js:4]: Move it to the constants file/folder

File: src/createServer.js [hunk 2]
```diff
@@ -1,3 +1,59 @@
...
+
+    if (!targetCase) {
+      errors.push({ message: `"toCase" query param is required. ${correctRequestText}` });
+    } else {
+      const upperCaseTargetCase = targetCase.toUpperCase();
+
+      if (!supportedCases.includes(upperCaseTargetCase)) {
+        errors.push({ message: `This case is not supported. ${availableCases}` });
+      }
+    }
```
Comment (BudnikOleksii) [inline src/createServer.js:33]: - Let's create a helper that will check for errors.
- put all error messages into a constant
- Available cases are in the `supportedCases` constant, use them.
- avoid else statements

File: src/createServer.js [hunk 3]
```diff
@@ -1,3 +1,59 @@
...
+
+    const response = {
+      status: errors.length > 0 ? 400 : 200,
+      headers: { 'Content-Type': 'application/json' },
+      body: errors.length > 0 ? { errors } : {},
+    };
+
+    if (errors.length === 0) {
```
Comment (BudnikOleksii) [inline src/createServer.js:41]:
```suggestion
    if (!errors.length) {
```

File: src/createServer.js [hunk 4]
```diff
@@ -1,3 +1,59 @@
...
+      const result = convertToCase(convertedText, targetCase.toUpperCase());
+
+      response.body = {
+        originalText: convertedText,
+        targetCase: targetCase.toUpperCase(),
+        convertedText: result.convertedText,
```
Comment (BudnikOleksii) [inline src/createServer.js:47]: - originalText is convertedText? 🤯 
- `targetCase.toUpperCase()` you do it twice

File: src/createServer.js [hunk 5]
```diff
@@ -1,3 +1,59 @@
...
+        originalCase: result.originalCase,
+      };
+    }
+
+    res.writeHead(response.status, response.headers);
+    res.end(JSON.stringify(response.body));
```
Comment (BudnikOleksii) [inline src/createServer.js:53]: Also, you need to add statusText:
- If something is not correct, you should respond with 400 status, Bad request statusText and the next payload
- You should respond with status 200 and OK status text.

**Review by BudnikOleksii — CHANGES_REQUESTED** [review]
Good work 👍 
Left some comments. Check them and improve your solution

File: src/createServer.js [hunk 6]
```diff
@@ -1,3 +1,60 @@
...
+
+    if (!targetCase) {
+      errors.push({ message: `"toCase" query param is required. ${correctRequestText}` });
+    } else {
+      const upperCaseTargetCase = targetCase.toUpperCase();
+
+      if (!supportedCases.includes(upperCaseTargetCase)) {
+        errors.push({ message: `This case is not supported. ${availableCases}` });
+      }
+    }
```
Comment (Moroz-Dmytro) [inline src/createServer.js:32]: Move this logic to separate module

File: src/createServer.js [hunk 7]
```diff
@@ -1,3 +1,60 @@
...
+    const path = parsedUrl.pathname;
+    const query = parsedUrl.searchParams;
+
+    const convertedText = path.slice(1);
+    const targetCase = query.get('toCase');
+
+    const errors = [];
+
+    const correctRequestText = 'Correct request is: "/<TEXT_TO_CONVERT>?toCase=<CASE_NAME>".';
+    const availableCases = 'Available cases: SNAKE, KEBAB, CAMEL, PASCAL, UPPER.';
```
Comment (Moroz-Dmytro) [inline src/createServer.js:18]: Use your constant with available cases and `join` here

File: src/createServer.js [hunk 8]
```diff
@@ -1,3 +1,60 @@
...
+      if (!supportedCases.includes(upperCaseTargetCase)) {
+        errors.push({ message: `This case is not supported. ${availableCases}` });
+      }
+    }
+
+    const response = {
+      status: errors.length > 0 ? 400 : 200,
+      statusText: errors.length > 0 ? 'Bad Request' : 'OK',
+      headers: { 'Content-Type': 'application/json' },
+      body: errors.length > 0 ? { errors } : {},
```
Comment (Moroz-Dmytro) [inline src/createServer.js:38]:
```suggestion
      status: errors.length ? 400 : 200,
      statusText: errors.length ? 'Bad Request' : 'OK',
      headers: { 'Content-Type': 'application/json' },
      body: errors.length ? { errors } : {},
```

File: src/createServer.js [hunk 9]
```diff
@@ -1,3 +1,60 @@
...
+    if (!errors.length) {
+      const upperCaseTargetCase = targetCase.toUpperCase();
+      const result = convertToCase(convertedText, upperCaseTargetCase);
+
+      response.body = {
+        originalText: convertedText,
+        targetCase: upperCaseTargetCase,
+        convertedText: result.convertedText,
+        originalCase: result.originalCase,
+      };
```
Comment (Moroz-Dmytro) [inline src/createServer.js:50]:
```suggestion
      const { convertedText, originalCase } = convertToCase(convertedText, upperCaseTargetCase);

      response.body = {
        originalText: convertedText,
        targetCase: upperCaseTargetCase,
        convertedText,
        originalCase,
      };
```

File: src/createServer.js [hunk 10]
```diff
@@ -1,3 +1,60 @@
...
+const { convertToCase } = require('./convertToCase/convertToCase');
+const { supportedCases } = require('./constants');
+
+const createServer = () => {
+  const server = http.createServer((req, res) => {
+    const parsedUrl = new URL(req.url, `http://${req.headers.host}`);
+    const path = parsedUrl.pathname;
+    const query = parsedUrl.searchParams;
+
+    const convertedText = path.slice(1);
```
Comment (Moroz-Dmytro) [inline src/createServer.js:12]:
```suggestion
    const originalText = path.slice(1);
```

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]

File: src/createServer.js [hunk 11]
```diff
@@ -1,3 +1,42 @@
...
+const { convertToCase } = require('./convertToCase/convertToCase');
+const { validateRequest } = require('./requestValidation');
+
+const createServer = () => {
+  const server = http.createServer((req, res) => {
+    const parsedUrl = new URL(req.url, `http://${req.headers.host}`);
+    const path = parsedUrl.pathname;
+    const query = parsedUrl.searchParams;
+    const originalText = path.slice(1);
+    const targetCase = query.get('toCase');
```
Comment (Moroz-Dmytro) [inline src/createServer.js:12]:
```suggestion
    const originalText = parsedUrl.pathname.slice(1);
    const targetCase = parsedUrl.searchParams.get('toCase');
```

File: src/createServer.js [hunk 12]
```diff
@@ -1,3 +1,42 @@
...
+    if (!errors.length) {
+      const upperCaseTargetCase = targetCase.toUpperCase();
+      const { convertedText: converted, originalCase } = convertToCase(originalText, upperCaseTargetCase);
+
+      response.body = {
+        originalText,
+        targetCase: upperCaseTargetCase,
+        convertedText: converted,
+        originalCase,
+      };
```
Comment (Moroz-Dmytro) [inline src/createServer.js:32]:
```suggestion
      const { convertedText, originalCase } = convertToCase(originalText, upperCaseTargetCase);

      response.body = {
        originalText,
        targetCase: upperCaseTargetCase,
        convertedText,
        originalCase,
      };
```

File: src/createServer.js [hunk 13]
```diff
@@ -1,3 +1,42 @@
...
+
+    const response = {
+      status: errors.length ? 400 : 200,
+      statusText: errors.length ? 'Bad Request' : 'OK',
+      headers: { 'Content-Type': 'application/json' },
+      body: errors.length ? { errors } : {},
+    };
+
+    if (!errors.length) {
+      const upperCaseTargetCase = targetCase.toUpperCase();
```
Comment (Moroz-Dmytro) [inline src/createServer.js:24]: Why need to do this? It's not needed

File: src/requestValidation.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable max-len */
+const { supportedCases } = require('./constants');
+
+function validateRequest(convertedText, targetCase) {
+  const errors = [];
+  const correctRequestText = 'Correct request is: "/<TEXT_TO_CONVERT>?toCase=<CASE_NAME>".';
+  const availableCases = `Available cases: ${supportedCases.join(', ')}.`;
```
Comment (Moroz-Dmytro) [inline src/requestValidation.js:7]: Move this to constants file and name them in UPPER_CASE

File: src/requestValidation.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+  if (!convertedText) {
+    errors.push({ message: `Text to convert is required. ${correctRequestText}` });
+  }
+
+  if (!targetCase) {
+    errors.push({ message: `"toCase" query param is required. ${correctRequestText}` });
+  } else {
+    const upperCaseTargetCase = targetCase.toUpperCase();
```
Comment (Moroz-Dmytro) [inline src/requestValidation.js:16]: I guess we don't need to do this

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]

File: src/createServer.js (hunk 13 above)
Comment (olegpaunov) [reply src/createServer.js:24]: Do you mean about the variable?

Comment (olegpaunov) [reply src/createServer.js:24]: why do we need condition ? or variable?

File: src/constants.js [hunk 16]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable max-len */
+const supportedCases = ['SNAKE', 'KEBAB', 'CAMEL', 'PASCAL', 'UPPER'];
+const correctRequestText = 'Correct request is: "/<TEXT_TO_CONVERT>?toCase=<CASE_NAME>".';
+const availableCases = `Available cases: ${supportedCases.join(', ')}.`;
```
Comment (Moroz-Dmytro) [inline src/constants.js:4]: All these constants should be named in UPPER_CASE

**Review by Moroz-Dmytro — APPROVED** [review]

**Pull request closed without merging**